x/distribution/types/dec_coin.go

```diff
@@ -0,0 +1,141 @@
+package types
+
+import (
+	"strings"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// Coins which can have additional decimal points
+type DecCoin struct {
+	Denom  string  `json:"denom"`
+	Amount sdk.Dec `json:"amount"`
+}
+
+func NewDecCoin(coin sdk.Coin) DecCoin {
+	return DecCoin{
+		Denom:  coin.Denom,
+		Amount: sdk.NewDecFromInt(coin.Amount),
+	}
+}
+
+// Adds amounts of two coins with same denom
+func (coin DecCoin) Plus(coinB DecCoin) DecCoin {
+	if !(coin.Denom == coinB.Denom) {
+		return coin
+	}
+	return DecCoin{coin.Denom, coin.Amount.Add(coinB.Amount)}
+}
+
+// Subtracts amounts of two coins with same denom
+func (coin DecCoin) Minus(coinB DecCoin) DecCoin {
+	if !(coin.Denom == coinB.Denom) {
+		return coin
+	}
+	return DecCoin{coin.Denom, coin.Amount.Sub(coinB.Amount)}
+}
+
+// return the decimal coins with trunctated decimals
+func (coin DecCoin) TruncateDecimal() sdk.Coin {
+	return sdk.NewCoin(coin.Denom, coin.Amount.TruncateInt())
+}
+
+//_______________________________________________________________________
+
+// coins with decimal
+type DecCoins []DecCoin
+
+func NewDecCoins(coins sdk.Coins) DecCoins {
+	dcs := make(DecCoins, len(coins))
+	for i, coin := range coins {
+		dcs[i] = NewDecCoin(coin)
+	}
+	return dcs
+}
+
+// return the coins with trunctated decimals
+func (coins DecCoins) TruncateDecimal() sdk.Coins {
+	out := make(sdk.Coins, len(coins))
+	for i, coin := range coins {
+		out[i] = coin.TruncateDecimal()
+	}
+	return out
+}
+
+// Plus combines two sets of coins
+// CONTRACT: Plus will never return Coins where one Coin has a 0 amount.
+func (coins DecCoins) Plus(coinsB DecCoins) DecCoins {
+	sum := ([]DecCoin)(nil)
```

Comment:
Is this the same as `make([]DecCoin, nil)` - can we use the alias `DecCoins` type?

Also, we know that the size of the added coin array will be at least the maximum of `len(coins)`, `len(coinsB)`, so we might as well allocate that much

Reply:
good idea, again, just copied code from `types/coins` - also the max length would actually be `len(coins) + len(coinsB)` if they held all different denoms.

Reply:
... on second glance I think that these functions should be restructured in a new PR which also addressed the same for `types/coins` - leaving this for now

Reply:
opened an issue https://github.com/cosmos/cosmos-sdk/issues/2443